feat: Adding a relay bot for matrix<>qaul bridge

Harshil-Jani opened the pull request:
Raising the PR so we can do smooth code review for the code which we will be updating into the matrix-bridge branch.
As per the current state state
1. Run the bridge client `cargo run --bin=qaul-matrix-bridge`
2. Invite user `@qaul-bot` to any of the matrix rooms.
3. In message send `!qaul` and it will respond.

Harshil-Jani committed:
feat: Adding a relay bot for matrix<>qaul bridge

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

MathJud commented on rust/clients/bridge/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+[package]
+name = "bridge"
```
The name should be more explanatory, e.g. `qaul-bridge` or `qaul-matrix-bridge`. IMHO `qaul-matrix-bridge` would be best, as it is the most self-explanatory.

MathJud commented on rust/clients/bridge/README.md:
```diff
@@ -0,0 +1,86 @@
+# qaul-cli
```
Please write a specific README for the binary, that explains what the binary does and how to start it.

MathJud commented on rust/clients/bridge/src/relay_bot.rs:
```diff
@@ -0,0 +1,78 @@
+use matrix_sdk::{
```
Please add the license and a short description to every new file.
See e.g. `chat.rs` for an example.

MathJud commented on rust/clients/bridge/src/relay_bot.rs:
```diff
@@ -0,0 +1,78 @@
+    password: &str,
+) -> Result<(), matrix_sdk::Error> {
+    // the location for `JsonStore` to save files to
+    let mut home = dirs::home_dir().expect("no home directory found");
```
We should create a matrix specific `matrix.yml` config file, which is saved in qaul's default storage location, for that it is working as everything else in qaul.
The location to save the file to, can be asked from libqaul via the RPC debug>StoragePathRequest message. (see `rust/libqaul/src/rpc/debug.proto`)
If there is more matrix-bridge related stuff, i suggest to create an extra `matrix` folder in the qaul storage directory and use it as the storage directory for all matrix related stuff.
For saving and reading the configuration libqaul uses the crate `config`. (see `rust/libqaul/src/storage/configuration`.)

MathJud commented on rust/clients/bridge/src/main.rs:
```diff
@@ -0,0 +1,104 @@
+    let mut stdin = io::BufReader::new(io::stdin()).lines();
+
+    // connect the matrix bot with the qaul-cli
+    relay_bot::connect();
```
please handle the returned result.

MathJud reviewed (commented):
congrats to this first successful integration of the matrix-sdk crate into qaul!
I added a few comments on how to integrate it further into the qaul system.

This will include the configuration system and the use of the RPC messaging system with libqaul:

* check libqaul events using the RPC system
* save the matrix stuff into the same directory as libqaul does, to keep everything in one place.

Harshil-Jani committed:
Connected Matrix World with Qaul World

By this commit, We have achieved a very initial milestone of connecting
the Matrix with Qaul. Currently, In Qaul we have a separate thread
running which listens for any activity into the Matrix room and
forwards the messagess into the Qaul feel list.

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
Adding a README for qaul-matrix-bridge

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani commented on rust/clients/bridge/src/relay_bot.rs:
```diff
@@ -0,0 +1,120 @@
+            }
+
+            if msg_body.contains("!users-list") {
+                // Here I need to trigger the RPC call from users.rs file.
```
@MathJud This is the place where I am interested in calling the rpc function from `users.rs`. 
Use case is that. In matrix room, a user sends a message `!users-list`. This should return the data which we will format with better strings. 

If we are successful to do just this, Next we can call the rpc function of `feed.rs` and that should allow us to send any messages which we write in qaul world to be sent over to the matrix room. And this should be enough to complete our first milestone.

MathJud replied:
There are several ways to do that.

1) The bridge client queries regularly libqaul via RPC for a list of users, e.g. all 10 seconds. This list is stored in a way to be accessed by different threads and functions (see the state storage object https://crates.io/crates/state )

2) Create an asynchronous event system, where you pause your function until the next user list arrived.

As we anyway will need the user information, and also will need to query several other information regularly I would go with solution 1. It is also much simpler to implement.

MathJud replied:
Which users do you want to deliver here?

a) all known users, independent if they are online or not (this is an ever growing list of users)
b) the users that are online at this time.

Harshil-Jani replied:
> Which users do you want to deliver here?
a) all known users, independent if they are online or not (this is an ever growing list of users)
b) the users that are online at this time.

It's (a)

Harshil-Jani committed:
Completed Version 0 goals

By this commit, The integration is completed for two way communications
and basic functionality where from the qaul you can send Hii into the
matrix room and from matrix room you can send Hii into the qaul. To add
more, We have two commands running on matrix-side which are !qaul for
acknowledgement of connection and !users-list for getting list of users
in qaul on matrix.

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
Get users in matrix room inside qaul

This commit makes a logic of seeing the users connected in qaul-room to
make sure that if rooms are public only people we know and want are
joined.

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
Creating the configuration file

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
Organize the configuration within the bridge binary

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
adding copyright to configuration

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
Make bridge more reliable

There was a problem when we run `feed list` and all the messages were
sent to matrix which is not correct. This commit solves it while
checking against the last_index from the configuration file.

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
Save the configuration when updated

With this commit, The bug of saving the configuration file when it gets
changed is fixed. For this, The MatrixConfiguration::save() method has
been written which takes a configuration and serializes it in YAML and
saves the file.

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

MathJud committed:
libqaul RPC: return request_id for group module

MathJud committed:
updated rust protobuf files

MathJud committed:
Fixed: The first message in a new group now always starts at 0

Harshil-Jani committed:
Group Conversation Feature

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
Cleaning up the code

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
Adding help command

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
Acknowledging the qaul user name when the message arrives in matrix

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
Only Allow Admins to Invite

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
Allowing multiple users in a matrix room, Removing the users from Matrix

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
Adding group-info on the matrix menu

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
Disable CLI for server compatibility

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
Acknowledge which qaul user was invited in matrix room

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
Layouting Qaul Menu which can be easily extended in future as required.

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
Template for particular MessageType from Matrix.

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
File sending from Matrix to Qaul

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
Qaul to Matrix file sharing

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
Adding Peer ID In group information

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
Adding CLI back temporarily

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

Harshil-Jani committed:
Fixing User Discovery Bug using RwLock on states

Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>
